packages/tangle/filter_test.go

```diff
@@ -51,14 +78,102 @@ func TestPowFilter_Filter(t *testing.T) {
 		require.NoError(t, err)
 		require.GreaterOrEqual(t, zeroes, testDifficulty)
 
-		m.On("Accept", msgPOWBytes, testPeer)
-		filter.Filter(msgPOWBytes, testPeer)
+		m.On("Accept", msgPOW, testPeer)
+		filter.Filter(msgPOW, testPeer)
 	})
 
 	m.AssertExpectations(t)
 }
 
+func TestAPowFilter_Filter(t *testing.T) {
+	filter := NewPowFilter(testWorker, testDifficulty)
+	pow.ApowWindow = testWindow
+	pow.AdaptiveRate = testRate
+
+	// set callbacks
+	m := &callbackMock{}
+	filter.OnAccept(m.Accept)
+	filter.OnReject(m.Reject)
+
+	msg := newTestAPOWMessage(time.Now())
+	msgBytes := msg.Bytes()
+	nonce, err := testWorker.Mine(context.Background(), msgBytes[:len(msgBytes)-len(msg.Signature())-pow.NonceBytes], testDifficulty)
+	require.NoError(t, err)
+
+	msg.nonce = nonce
+	msg.bytes = nil
+
+	// accpet since apow difficulty should be >= diffficulty
+	m.On("Accept", msg, testPeer)
+	filter.Filter(msg, testPeer)
+
+	// loop to force the number of zeros to be == testDifficulty
+	zeros := 0
+	for zeros != testDifficulty {
```

Comment:
`Mine` should output the valid nonce with the given `testDifficulty`, why do we need this loop?

Reply:
Because mine could also output a nonce with a greater number of zeroes and we need to test the case of rejecting a nonce when the provided difficulty is not increased accordingly to the aPoW.